internal/commands/show/show.go

```diff
@@ -0,0 +1,217 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package show
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"io"
+	"os"
+
+	"github.com/github/smimesign/ietf-cms/protocol"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	v02 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/sigstore/gitsign/internal/config"
+	"github.com/sigstore/gitsign/pkg/predicate"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/spf13/cobra"
+)
+
+const (
+	predicateType = "gitsign.sigstore.dev/predicate/git/v0.1"
+)
+
+type options struct {
+	FlagRemote string
+}
+
+func (o *options) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.FlagRemote, "remote", "r", "origin", "make a signature")
+}
+
+func (o *options) Run(w io.Writer, args []string) error {
+	repo, err := git.PlainOpenWithOptions(".", &git.PlainOpenOptions{
+		DetectDotGit: true,
+	})
+	if err != nil {
+		return err
+	}
+	revision := "HEAD"
+	if len(args) > 0 {
+		revision = args[0]
+	}
+
+	out, err := statement(repo, o.FlagRemote, revision)
+	if err != nil {
+		return err
+	}
+
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	if err := enc.Encode(out); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func statement(repo *git.Repository, remote, revision string) (*in_toto.Statement, error) {
+	hash, err := repo.ResolveRevision(plumbing.Revision(revision))
+	if err != nil {
+		return nil, err
+	}
+	commit, err := repo.CommitObject(*hash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract parent hashes
+	parents := make([]string, 0, len(commit.ParentHashes))
+	for _, p := range commit.ParentHashes {
+		if !p.IsZero() {
+			parents = append(parents, p.String())
+		}
+	}
+
+	// Build initial predicate from the commit.
+	predicate := &predicate.GitCommit{
+		Commit: &predicate.Commit{
+			Tree:    commit.TreeHash.String(),
+			Parents: parents,
+			Author: &predicate.Author{
+				Name:  commit.Author.Name,
+				Email: commit.Author.Email,
+				Date:  commit.Author.When,
+			},
+			Committer: &predicate.Author{
+				Name:  commit.Committer.Name,
+				Email: commit.Committer.Email,
+				Date:  commit.Committer.When,
+			},
+			Message: commit.Message,
+		},
+		Signature: commit.PGPSignature,
+	}
+
+	// If we can a PEM encoded signature, try and extract certificate details.
+	pem, _ := pem.Decode([]byte(commit.PGPSignature))
```

Comment:
Maybe better not to ignore `err` checking and `pem.Bytes > 0` for more strict error checking?

Reply:
Actually not an error! It's the left over bytes found during PEM decoding. https://pkg.go.dev/encoding/pem#Decode